src/dotnet-grpc/Commands/AddFileCommand.cs

```diff
@@ -0,0 +1,79 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System.CommandLine;
+using System.CommandLine.Invocation;
+using System.IO;
+using Grpc.Dotnet.Cli.Internal;
+using Grpc.Dotnet.Cli.Options;
+
+namespace Grpc.Dotnet.Cli.Commands
+{
+    internal class AddFileCommand : CommandBase
+    {
+        public static Command Create()
+        {
+            var command = new Command(
+                name: "file",
+                description: "Add protobuf file reference(s) to the gRPC project.",
+                argument: new Argument<string[]>
+                {
+                    Name = "files",
+                    Description = "The protobuf file reference(s). These can be a path to glob for local protobuf file(s).",
+                });
+
+            command.AddOption(CommonOptions.ProjectOption());
+            command.AddOption(CommonOptions.ServiceOption());
+            command.AddOption(CommonOptions.AdditionalImportDirsOption());
+            command.AddOption(CommonOptions.AccessOption());
+
+            command.Handler = CommandHandler.Create<IConsole, FileInfo, Services, Access, string, string[]>(new AddFileCommand().AddFile);
```

Comment:
You could add an `Initialize(IConsole, FileInfo)` to CommandBase. It would handle setting the console and project file. It would be called here. You would need to move the try/catch to return 1 here as well.

Alternatively you could do the initialization in the command constructors. That would allow them to be marked as not null.

The downside would be specifying an anonymous delegate rather than passing a method directly like you're currently doing.

What do you think? There are pros and cons.

Reply:
I like the constructor approach.